[Docs] Beta TestNet - Supplier Cheatsheet (#947)
pokt-network/poktroll: docusaurus/docs/operate/quickstart/supplier_cheatsheet.md

---
sidebar_position: 6
title: Supplier (RelayMiner) Cheat Sheet
---

## Supplier Cheat Sheet <!-- omit in toc -->

This guide provides quick reference commands for setting up a **Supplier** and
running a **RelayMiner** on Pocket Network.

:::warning

These instructions are intended to run on a Linux machine.

> TODO_TECHDEBT(@olshansky): Adapt the instructions to be macOS friendly.

Author: Updating TODOs like so:

```
TODO_TECHDEBT(@olshansky): Adapt instructions to be macOS friendly in order to
streamline development and reduce friction for any new potential contributor.
```

North star: If anyone TRIES to run things, I want to make sure that in 2 mins or less, no one hits issues

  - [How do I query for all existing onchain Suppliers?](#how-do-i-query-for-all-existing-onchain-suppliers)

:::note

For detailed instructions, troubleshooting, and observability setup, see the [Supplier Walkthrough](./../run_a_node/supplier_walkthrough.md).

:::

## Pre-Requisites

1. Make sure to [install the `poktrolld` CLI](../user_guide/install.md).
2. Make sure you know how to [create and fund a new account](../user_guide/create-new-wallet.md).
3. You have either [staked a new `service` or found an existing one](./service_cheatsheet.md).
4. `[Optional]` You can run things locally or have dedicated long-running hardware. See the [Docker Compose Cheat Sheet](./docker_compose_debian_cheatsheet#deploy-your-server) if you're interested in the latter.

### Context

This document is a cheat sheet to get you quickly started with two things:

1. Staking an onchain `Supplier`
2. Deploying an offchain `RelayMiner`

By the end of it, you should be able to serve Relays onchain.

## Account Setup

### Create and fund the `Supplier` account

Create a new key pair for the `Supplier`

```bash
poktrolld keys add supplier

# Optionally, to avoid entering the password each time:

> # poktrolld keys add supplier --keyring-backend test

Author: Add a warning + note at the top

```

:::tip

You can set the `--keyring-backend` flag to `test` to avoid entering the password
each time.

Learn more about [cosmos keyring backends here](https://docs.cosmos.network/v0.46/run-node/keyring.html).

:::

### Prepare your environment

For convenience, we're setting several environment variables to streamline
the process of interacting with the Shannon network:

We recommend you put these in your `~/.bashrc` file:

```bash

> export NODE="https://shannon-testnet-grove-rpc.beta.poktroll.com"

Reviewer: Maybe also add a note about where to find the config files and that this modifies the node field of the client config.

Also note, the poktrolld "home" can be discovered with:

```bash
poktrolld config home
```

> export NODE="https://shannon-testnet-grove-rpc.beta.poktroll.com"

Author: @bryanchriswhite Thank you for these suggestions! #TIL

To limit the scope of this PR, I created #1011

> export NODE_FLAGS="--node=https://shannon-testnet-grove-rpc.beta.poktroll.com"

Reviewer: Given the above, we shouldn't need this.

> export NODE_FLAGS="--node=https://shannon-testnet-grove-rpc.beta.poktroll.com"

Author: https://github.com/pokt-network/poktroll/pull/947#discussion_r1887589442

export TX_PARAM_FLAGS="--gas=auto --gas-prices=1upokt --gas-adjustment=1.5 --chain-id=pocket-beta --yes"
export SUPPLIER_ADDR=$(poktrolld keys show supplier -a)

# Optionally, to avoid entering the password each time:
# export SUPPLIER_ADDR=$(poktrolld keys show supplier -a --keyring-backend test
```

:::tip

You can put the above in a special `~/.poktrollrc` and add `source ~/.poktrollrc` to

> your `~/.profile` file for a cleaner organization.

Reviewer: :sunglasses: Nice!

> You can put the above in a special `~/.poktrollrc` and add `source ~/.poktrollrc` to
> your `~/.profile` file for a cleaner organization.

Reviewer: ```suggestion
As an alternative to appending directly to `~/.bashrc`, you can put the above in a special `~/.poktrollrc` and add `source ~/.poktrollrc` to
your `~/.profile` (or `~/.bashrc`) file for a cleaner organization.
```

:::

## Supplier Configuration

### Fund the Supplier account

Run the following command to get the `Supplier`:

```bash
echo "Supplier address: $SUPPLIER_ADDR"
```

Then use the [Shannon Beta TestNet faucet](https://faucet.beta.testnet.pokt.network/) to fund the account.

Afterwards, you can query the balance using the following command:

```bash

> poktrolld query bank balances $SUPPLIER_ADDR $NODE_FLAGS

Reviewer: ```suggestion
poktrolld query bank balances $SUPPLIER_ADDR
```

Author: Resolving as wont-do and delegating to #1011

```

:::tip

You can find all the explorers, faucets and tools at the [tools page](../../explore/tools.md).

:::

### Stake the Supplier

:::info

For an in-depth look at how to stake a supplier, see the [Supplier configuration docs](./../configs/supplier_staking_config.md).

The example below is a very quick and simple way to get you started by staking for
Pocket Network's Morse service on Shannon using a public RPC endpoint provided by

> [Liquify](https://liquify.com/).

Reviewer: I find this sentence super dense. It's wonderful :star: , but I had to re-read it like 3 times to fully unpack it in my head. :sweat_smile:

Author: Dense is not bad.

Wonderful is great.

Needing to re-read 3 times is bad.

Unless you're a reviewer or getting paid to read it, no one else will.

:::

Retrieve your external IP address:

```bash

> EXTERNAL_IP=$(curl -4 ifconfig.me/ip)

Reviewer: :fire:

```

Choose a port that'll be publicly accessible from the internet (e.g. `8545`)

```bash

> sudo ufw allow 8545/tcp

Reviewer: Consider adding a tip or something with a note about the fact that this only applies if your operating system uses  ufw and a link (e.g. [arch wiki - Uncomplicated Firewall](https://wiki.archlinux.org/title/Uncomplicated_Firewall)).

```

Create a Supplier stake configuration file:

```bash

> cat <<EOF > /tmp/stake_supplier_config.yaml

Reviewer: (Yes, this actually works :sunglasses:)
 
```suggestion
cat <<🚀 > /tmp/stake_supplier_config.yaml
```

owner_address: $SUPPLIER_ADDR
operator_address: $SUPPLIER_ADDR
stake_amount: 1000069upokt
default_rev_share_percent:
  $SUPPLIER_ADDR: 100
services:
  - service_id: "morse"
    endpoints:
      - publicly_exposed_url: http://$EXTERNAL_IP:8545
        rpc_type: JSON_RPC

> EOF

Reviewer: ```suggestion
🚀
```

```

And run the following command to stake the `Supplier`:

```bash

> poktrolld tx supplier stake-supplier --config /tmp/stake_supplier_config.yaml --from=$SUPPLIER_ADDR $TX_PARAM_FLAGS $NODE_FLAGS

Reviewer: ```suggestion
poktrolld tx supplier stake-supplier --config /tmp/stake_supplier_config.yaml --from=$SUPPLIER_ADDR $TX_PARAM_FLAGS
```

Author: Resolving as wont-do and delegating to #1011

> # Optionally, to avoid entering the password each time:
> # poktrolld tx supplier stake-supplier --config /tmp/stake_supplier_config.yaml --from=$SUPPLIER_ADDR $TX_PARAM_FLAGS $NODE_FLAGS --keyring-backend test

Reviewer: (same feedback as above)

```suggestion
# Optionally, to avoid entering the password each time:
# poktrolld tx supplier stake-supplier --config /tmp/stake_supplier_config.yaml --from=$SUPPLIER_ADDR $TX_PARAM_FLAGS --keyring-backend test
```

Author: Resolving as wont-do and delegating to #1011

```

After about a minute, you can check the `Supplier`'s status like so:

```bash

> poktrolld query supplier show-supplier $SUPPLIER_ADDR $NODE_FLAGS

Reviewer: ```suggestion
poktrolld query supplier show-supplier $SUPPLIER_ADDR
```

```

## RelayMiner Configuration

### Configure the RelayMiner

```bash

> cat <<EOF > /tmp/relayminer_config.yaml

Reviewer: ```suggestion
cat <<🚀 > /tmp/relayminer_config.yaml
```

default_signing_key_names:
  - supplier
smt_store_path: /home/pocket/.poktroll/smt
pocket_node:
  query_node_rpc_url: https://shannon-testnet-grove-rpc.beta.poktroll.com
  query_node_grpc_url: https://shannon-testnet-grove-grpc.beta.poktroll.com:443
  tx_node_rpc_url: https://shannon-testnet-grove-rpc.beta.poktroll.com
suppliers:
  - service_id: "morse"
    service_config:
      backend_url: "https://pocket-rpc.liquify.com"
      publicly_exposed_endpoints:
        - $EXTERNAL_IP
    listen_url: http://0.0.0.0:8545
metrics:
  enabled: false
  addr: :9090
pprof:
  enabled: false
  addr: :6060

> EOF

Reviewer: ```suggestion
🚀
```

```

### Start the RelayMiner

```bash
poktrolld \
    relayminer \
    --grpc-insecure=false \
    --log_level=debug \
    --config=/tmp/relayminer_config.yaml \
    # --keyring-backend=test
```

### Secure vs Non-Secure `query_node_grpc_url`

> In `/tmp/relayminer_config.yaml`, you'll see that we specify an endpoint with TLS
> for `query_node_grpc_url`.

Reviewer: ```suggestion
In `/tmp/relayminer_config.yaml`, you'll see that we specify an endpoint for
`query_node_grpc_url` which is TLS terminated.
```

If `grpc-insecure=true` then it **MUST** be an HTTP port, no TLS.

> The Grove team exposed one such endpoint on one of our validators for Beta Testnet at `http://149.28.34.68:9090`.

Reviewer 2: Let's remove the IP as we don't officially maintain this endpoint. We do advertise and maintain the HTTPs endpoint.

Something like:

```
### Secure vs Non-Secure `query_node_grpc_url`

In `/tmp/relayminer_config.yaml`, you'll see that we specify an endpoint with TLS
for `query_node_grpc_url`.

If `grpc-insecure=true` then it **MUST** be an HTTP port, no TLS.

To validate this, you can use `grpcurl` with the `-plaintext` flag, indicating no TLS encryption. For example: 

```
grpcurl -plaintext <host>:<port> list
```

If `grpc-insecure=false`, then it **MUST** be an HTTPS port, with TLS.

The Grove team exposed one such endpoint on one of our validators for Beta Testnet at `https://shannon-testnet-grove-grpc.beta.poktroll.com:443`.
It can be validated with:

```
grpcurl shannon-testnet-grove-grpc.beta.poktroll.com:443 list
```

Note that no `-plaintext` flag is required here, indicating TLS encryption is enabled.

:::tip

You can replace both `http` and `https` with `tcp` and it should work the same.

:::
```

> It can be validated with `grpcurl -plaintext 149.28.34.68:9090`; note that the `-plaintext` flag meaning no TLS encryption.

Reviewer 2: ```suggestion
It can be validated with `grpcurl -plaintext 149.28.34.68:9090 list`; note that the `-plaintext` flag meaning no TLS encryption.
```

If `grpc-insecure=false`, then it **MUST** be an HTTPS port, with TLS.

The Grove team exposed one such endpoint on one of our validators for Beta Testnet at `https://shannon-testnet-grove-grpc.beta.poktroll.com:443`.

> It can be validated with `grpcurl https://shannon-testnet-grove-grpc.beta.poktroll.com:443`; note no `-plaintext` flag meaning no TLS encryption.

Reviewer 2: ```suggestion
It can be validated with `grpcurl shannon-testnet-grove-grpc.beta.poktroll.com:443 list`; note no `-plaintext` flag meaning no TLS encryption.
```

> It can be validated with `grpcurl https://shannon-testnet-grove-grpc.beta.poktroll.com:443`; note no `-plaintext` flag meaning no TLS encryption.

Reviewer 3: ```suggestion
It can be validated with `grpcurl https://shannon-testnet-grove-grpc.beta.poktroll.com:443`; note the absence of the `-plaintext` flag meaning no TLS encryption.
```

:::tip

You can replace both `http` and `https` with `tcp` and it should work the same.

:::

## Supplier FAQ

> ### What Supplier transactions are available?

Reviewer: Wdyt about "operations" instead of "transactions" to disambiguate?

```suggestion
### What Supplier operations are available?
```
